### PR TITLE
Proposed fix for: some ST boards fail to build `samples/sensor/sensor_shell`

### LINKED ISSUE #72910: some ST boards fail to build `samples/sensor/sensor_shell`
**Describe the bug**
`nucleo_h753zi` fails to build `samples/sensor/sensor_shell`, because the `vbat` driver depends on `adc3`, but `adc3` is not enabled in the devicetree.

**To Reproduce**
1. `west build -b nucleo_h753zi -p auto zephyr/samples/sensor/sensor_shell`

**Expected behavior**
Sample should compile

**Impact**
This is failing in #72833

**Logs and console output**
```
In file included from /home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/irq_multilevel.h:15,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/devicetree.h:20,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:12,
                 from /home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:7:
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:91:41: error: '__device_dts_ord_10' undeclared here (not in a function); did you mean '__device_dts_ord_50'?
   91 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/devicetree.h:4305:9: note: in expansion of macro 'COND_CODE_1'
 4305 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/toolchain/common.h:138:23: note: in expansion of macro '_DO_CONCAT'
  138 | #define _CONCAT(x, y) _DO_CONCAT(x, y)
      |                       ^~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:91:33: note: in expansion of macro '_CONCAT'
   91 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:228:37: note: in expansion of macro 'DEVICE_NAME_GET'
  228 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:245:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  245 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:121:24: note: in expansion of macro 'DEVICE_DT_GET'
  121 |                 .adc = DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst)),                           \
      |                        ^~~~~~~~~~~~~
/home/ycsin/zephyrproject/build/zephyr/include/generated/devicetree_generated.h:24582:48: note: in expansion of macro 'STM32_VBAT_DEFINE'
24582 | #define DT_FOREACH_OKAY_INST_st_stm32_vbat(fn) fn(0)
      |                                                ^~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_st_stm32_vbat'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:141:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
  141 | DT_INST_FOREACH_STATUS_OKAY(STM32_VBAT_DEFINE)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[74/166] Building C object zephyr/arch/arch/arm/core/cortex_m/CMakeFiles/arch__arm__core__cortex_m.dir/prep_c.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/ycsin/zephyrproject/build
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: 0.16.5-1
 - `main` branch

### PR DESCRIPTION
**Pull request by ycsin**
Fixes #72910

### CODE REVIEW

**Comment by ycsin** [comment]
The sample is failing for `stm32h735g_disco` as well in #72909, the driver patch here will sorta bury the issue as they will pass build-only test

File: drivers/sensor/st/stm32_vbat/stm32_vbat.c [hunk 1]
```diff
@@ -116,9 +121,13 @@ static int stm32_vbat_init(const struct device *dev)
 	return 0;
 }
 
+#define STM32_VBAT_GET_ADC_OR_NULL(inst)                                                           \
+	COND_CODE_1(DT_NODE_HAS_STATUS(DT_INST_IO_CHANNELS_CTLR(inst), okay),                      \
+		    (DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst))), (NULL))
```
Comment (erwango) [inline drivers/sensor/st/stm32_vbat/stm32_vbat.c:126]: What about
```suggestion
	IF_ENABLED(DT_NODE_HAS_STATUS_OKAY(DT_INST_IO_CHANNELS_CTLR(inst)),                      \
		               (DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst))))
```

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the fix @ycsin.
One suggestion otherwise LGTM

File: drivers/sensor/st/stm32_vbat/stm32_vbat.c [hunk 2]
```diff
@@ -101,6 +101,11 @@ static int stm32_vbat_init(const struct device *dev)
 
 	k_mutex_init(&data->mutex);
 
+	if (data->adc == NULL) {
```
Comment (bbilas) [inline drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: This shouldn't be needed as the "device_is_ready" function allows NULL to be passed, but I see there is a pointer dereference that could be a problem.

Comment (ycsin) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: I can do that but then there will be no difference between 'adc driver init failure' vs 'adc node not enabled in devicetree'

cc @erwango, any preference?

Comment (bbilas) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: Is it possible if `DT_HAS_ST_STM32_ADC_ENABLED` has been added?

Comment (ycsin) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: As mentioned in the commit,  that checks if we have at least an adc enabled in the DT, but not necessarily the one used by vbat

Comment (erwango) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: > DT_HAS_ST_STM32_ADC_ENABLED

Yes, another adc instance could be enabled, but not the one we need.

Comment (bbilas) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: > As mentioned in the commit, that checks if we have at least an adc enabled in the DT, but not necessarily the one used by vbat

Fair enough.

Comment (erwango) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:104]: > I can do that but then there will be no difference between 'adc driver init failure' vs 'adc node not enabled in devicetree'

Indeed, `'adc node not enabled in devicetree'` can be a tricky one, it's better if we're able to provide a clear error message on this specific issue.

File: drivers/sensor/st/stm32_vbat/stm32_vbat.c (hunk 1 above)
Comment (ycsin) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:126]: We do not have `DT_NODE_HAS_STATUS_OKAY` API

Comment (erwango) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:126]: I've seen the other PR indeed. Strange.. Thanks for having had a try anyway

**Review by erwango — APPROVED** [review]

File: drivers/sensor/st/stm32_vbat/stm32_vbat.c (hunk 1 above)
Comment (bbilas) [reply drivers/sensor/st/stm32_vbat/stm32_vbat.c:126]: https://github.com/zephyrproject-rtos/zephyr/blame/main/soc/nxp/rw/soc.c#L293 @decsny it looks like it won't work.

**Review by bbilas — APPROVED** [review]

**Pull request merged**